sdk/data/azcosmos/cosmos_example_test.go

```diff
@@ -0,0 +1,136 @@
+	// ===== 2. Creating a container =====
+
+	properties := CosmosContainerProperties{
+		Id: "aContainer",
+		PartitionKeyDefinition: PartitionKeyDefinition{
+			Paths: []string{"/id"},
+		},
+		IndexingPolicy: &IndexingPolicy{
+			IncludedPaths: []IncludedPath{
+				{Path: "/*"},
+			},
+			ExcludedPaths: []ExcludedPath{
+				{Path: "/\"_etag\"/?"},
+			},
+			Automatic:    true,
+			IndexingMode: IndexingModeConsistent,
+		},
+	}
+
+	throughput := NewManualThroughputProperties(400)
+
+	resp, _ := database.DatabaseProperties.Database.CreateContainer(ctx, properties, throughput, nil)
+	container := resp.ContainerProperties.Container
+
+	// ===== 2. Update container properties =====
+
+	resp, _ = container.Read(ctx, nil)
+
+	updatedProperties := CosmosContainerProperties{
+		Id: "aContainer",
+		PartitionKeyDefinition: PartitionKeyDefinition{
+			Paths: []string{"/id"},
+		},
+		ETag: "someEtag",
+		IndexingPolicy: &IndexingPolicy{
+			IncludedPaths: []IncludedPath{},
+			ExcludedPaths: []ExcludedPath{},
+			Automatic:     false,
+			IndexingMode:  IndexingModeNone,
+		},
+	}
+
+	// Replace container properties
+	resp, _ = container.Replace(ctx, updatedProperties, nil)
+
+	// Read the manual throughput property
+	throughputResponse, _ := container.ReadThroughput(ctx, nil)
+	_, _ = throughputResponse.ThroughputProperties.ManualThroughput()
+
+	// Replace manual throughput property
+	newScale := NewManualThroughputProperties(500)
+	_, _ = container.ReplaceThroughput(ctx, *newScale, nil)
+
+	// Container autoscale throughput
+	_, _ = throughputResponse.ThroughputProperties.AutoscaleMaxThroughput()
+
+	newScale = NewAutoscaleThroughputProperties(10000)
+	_, _ = container.ReplaceThroughput(ctx, *newScale, nil)
+
+	// Delete container
+	resp, _ = container.Delete(ctx, nil)
+
+	// ===== 3. Item CRUD =====
+
+	// Items in an Azure Cosmos container are uniquely identified by their id and partition key value.
+	pk, _ := NewPartitionKey("1")
+
+	item := map[string]string{
+		"id":    "1",
+		"value": "2",
+	}
+
+	// Create item.
+	itemResponse, _ := container.CreateItem(ctx, pk, item, nil)
+
+	// Read item.
+	itemResponse, _ = container.ReadItem(ctx, pk, "1", nil)
+
+	// Item response body
+	var itemResponseBody map[string]interface{}
+	err = json.Unmarshal(itemResponse.Value, &itemResponseBody)
+	log.Panic(err)
+
+	// ===== 4. Optimistic Concurrency =====
+
+	// Azure Cosmos DB supports optimistic concurrency control to prevent lost updates or deletes and detection of conflicting operations.
+	itemResponse, _ = container.UpsertItem(ctx, pk, item, &CosmosItemRequestOptions{EnableContentResponseOnWrite: true})
+
+	itemSessionToken := itemResponse.SessionToken()
+
+	itemResponse, _ = container.ReadItem(ctx, pk, "1", &CosmosItemRequestOptions{SessionToken: itemSessionToken})
+
+	// ===== 5. Etag PreConditionFail =====
+	itemEtag := itemResponse.ETag()
+
+	itemResponse, _ = container.ReadItem(ctx, pk, "1", &CosmosItemRequestOptions{IfMatchEtag: &itemEtag})
```

Comment:
The gist of Optimistic Concurrency is that you do:
1. ReadItem
2. Get Etag from read response
3. Modify the document value, change anything
4. Call ReplaceItem, passing IfMatchEtag
5. Check if the response was a failure and the error was StatusCode 412
6. If it was 412, repeat point 1, 2, 3, 4 until it succeeds

Reply:
I'll add